### PR TITLE
Improve fetching and caching job details

### PR DESCRIPTION
**Pull request by harshthakkar01**
#### Manual testing

```
2024-11-06 19:18:10,905 INFO: Printing job details --> [Job(id='5', name='test', required_nodes=None, job_state='RUNNING', duration=datetime.timedelta(days=365)), Job(id='6', name='test', required_nodes=None, job_state='PENDING', duration=None), Job(id='7', name='test', required_nodes=None, job_state='PENDING', duration=None), Job(id='8', name='test', required_nodes=None, job_state='PENDING', duration=None)]
2024-11-06 19:18:10,906 INFO: keep jobs --> {'7', '0', '8', '6', '5'}
```

#### Unit test
```
(virtualenv) harshthakkar@harshthakkar:~/github/workspace/hpc-toolkit$ pytest community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py
============================================================================================================================= test session starts ==============================================================================================================================
platform linux -- Python 3.11.9, pytest-8.3.3, pluggy-1.5.0
rootdir: /usr/local/google/home/harshthakkar/github/workspace/hpc-toolkit
configfile: .pytest.ini
plugins: mock-3.14.0
collected 32 items

community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py ................................                                                                                                                              [100%]

============================================================================================================================== 32 passed in 0.49s ==============================================================================================================================
```

### Submission Checklist

NOTE: Community submissions can take up to 2 weeks to be reviewed.

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 1]
```diff
@@ -68,6 +69,18 @@
 )
 
 
+KeepStates = frozenset(
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:72]: This belongs to utils.py

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 2]
```diff
@@ -99,6 +112,15 @@ def _find_dynamic_node_status() -> NodeStatus:
     return NodeStatus.unchanged  # don't touch dynamic nodes
 
 
+@lru_cache
+def get_keep_jobs():
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:116]: Why filter by `keep` state?

It should be a method on `Lookup`, otherwise cahce will be shared across multiple "Lookups", this would make for hard testing )

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 3]
```diff
@@ -99,6 +112,15 @@ def _find_dynamic_node_status() -> NodeStatus:
...
+    return {
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:117]: This belongs to util.py

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 4]
```diff
@@ -99,6 +112,15 @@ def _find_dynamic_node_status() -> NodeStatus:
     return NodeStatus.unchanged  # don't touch dynamic nodes
 
 
+@lru_cache
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:115]: It should return a collection of `Job`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 5]
```diff
@@ -99,6 +112,15 @@ def _find_dynamic_node_status() -> NodeStatus:
...
+def get_keep_jobs():
+    return {
+        str(job["job_id"])
+        for job in json.loads(run(f"{lookup().scontrol} show jobs --json").stdout)["jobs"]
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:119]: Did you consider to not use `--json`?

Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py]: updated.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py (hunk 4 above)
Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py]: ack.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py (hunk 2 above)
Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py]: Created a method in lookup to return all jobs details. Here we can filter by states as it was previously implemented.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 6]
```diff
@@ -1923,26 +1939,62 @@ def nodeset_map(self, hostnames: list):
             nodeset_map[self.node_nodeset_name(node)].append(node)
         return nodeset_map
 
+    @lru_cache
+    def get_jobs(self) -> List[Job]:
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1943]: `job` and `get_jobs` should return the same level of details.
Please create an internal function that can parse the the output format of `scontrol` and reuse it.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 7]
```diff
@@ -327,20 +327,11 @@ def ignore_err(e) -> bool:
 
 def sync_placement_groups():
     """Delete placement policies that are for jobs that have completed/terminated"""
-    keep_states = frozenset(
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:330]: Why did you move it? 
When I said that "something belongs to util" I've meant the `get_jobs` function.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 8]
```diff
@@ -1923,26 +1939,62 @@ def nodeset_map(self, hostnames: list):
...
-            seconds=int(seconds)
-        )
-        return Job(id=job_id, duration=job_duration)
+        namePattern = r"JobName=(\w+)"
+        matchName = re.search(namePattern, jobInfo)
+
+        if matchName:
+            name = matchName.group(1)
+
+        jobStatePattern = r"JobState=(\w+)"
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1985]:
```py
job_state = None
...
jobStatePattern = r"JobState=(\w+)"
matchJobState = re.search(jobStatePattern, jobInfo)

if matchJobState:
   job_state = matchJobState.group(1)
```

if match := re.search(r"JobState=(\w+)", jobInfo):
  job_state = match.group(1)
else:
  job_state = None

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 9]
```diff
@@ -1923,26 +1939,62 @@ def nodeset_map(self, hostnames: list):
...
+    def get_jobs(self) -> List[Job]:
+        res = run(f"{self.scontrol} show jobs", timeout=30)
+        all_jobs = [x.split() for x in res.stdout.split("\n\n")[:-1]]
+
+        res_jobs: List[Job] = []
+        for job in all_jobs:
+            for item in job:
+                key, value = item.split('=', 1) # Split at the first '='
+                if key == 'JobId':
+                    res_jobs.append(self.job(value))
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1952]: This wold cause multiple calls to `scontrol`, can't we extract information from returned result already?

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 6 above)
Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py]: done.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 9 above)
Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py]: Ack, moved the common logic to separate internal function.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 10]
```diff
@@ -68,6 +68,18 @@
 )
 
 
+KeepStates = frozenset(
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:71]: I wouldn't move it to global scope as it doesn't get used outside of the `sync_placement_groups`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 11]
```diff
@@ -327,20 +339,11 @@ def ignore_err(e) -> bool:
...
-        ]
-    )
 
     keep_jobs = {
-        str(job["job_id"])
-        for job in json.loads(run(f"{lookup().scontrol} show jobs --json").stdout)["jobs"]
-        if "job_state" in job and set(job["job_state"]) & keep_states
+        str(job.id)
+        for job in lookup().get_jobs()
+        if job.job_state and job.job_state in KeepStates
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:346]: `if job.job_state and job.job_state in KeepStates` -> `if job.job_state in KeepStates`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 12]
```diff
@@ -1928,26 +1932,58 @@ def nodeset_map(self, hostnames: list):
             nodeset_map[self.node_nodeset_name(node)].append(node)
         return nodeset_map
 
+    def _get_job_info(self, job_info, job_id=None) -> Optional[Job]:
+        """Extract job details"""
+        if job_id is None:
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1937]: It's not clear what the point of `job_id` here.
If `job_info` contains different `JobId` it would be just silently overwritten.  
My advice would be to either get rid of `job_id` argument, or add validation that `JobId` matches `job_id`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 13]
```diff
@@ -1928,26 +1932,58 @@ def nodeset_map(self, hostnames: list):
             nodeset_map[self.node_nodeset_name(node)].append(node)
         return nodeset_map
 
+    def _get_job_info(self, job_info, job_id=None) -> Optional[Job]:
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1935]: add type annotation to arguments.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 14]
```diff
@@ -1928,26 +1932,58 @@ def nodeset_map(self, hostnames: list):
...
+        """Extract job details"""
+        if job_id is None:
+            if match:= re.search(r"JobId=(\d+)", job_info):
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1938]: It's never the case that `JobId` is not present, should throw an exception if this happens.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 13 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1935]: Why is it `Optional` result?

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 15]
```diff
@@ -1928,26 +1932,58 @@ def nodeset_map(self, hostnames: list):
...
+            name = match.group(1)
+        else:
+            name = None
+
+        if match := re.search(r"JobState=(\w+)", job_info):
+            job_state = match.group(1)
+        else:
+            job_state = None
+
+        if match := re.search(r"ReqNodeList=(\w+)", job_info):
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1962]: Invalid regex, it'doesn't match "ReqNodeList=node-a,node-b"
Please also check if `ReqNodeList` can return "slurm hostlist" format, e.g. `node-[1-7,12,20-26]`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 13 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1935]: it would be nice to cover with unit tests

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 16]
```diff
@@ -1928,26 +1932,58 @@ def nodeset_map(self, hostnames: list):
...
-        jobInfo = run(f"{self.scontrol} show jobid {job_id}", check=False).stdout.rstrip()
-        if not jobInfo:
-            return None
+    def get_jobs(self) -> List[Job]:
+        res = run(f"{self.scontrol} show jobs", timeout=30)
+        all_jobs = [x for x in res.stdout.split("\n\n")[:-1]]
 
-        timePattern = r"TimeLimit=(?:(\d+)-)?(\d{2}):(\d{2}):(\d{2})"
-        match = re.search(timePattern, jobInfo)
+        res_jobs: List[Job] = []
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1974]: `return [self._parse_job_info(j) for b in all_jobs]`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 13 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1935]: nit. rename `_get_job_info` -> `_parse_job_info`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py (hunk 15 above)
Comment (harshthakkar01) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py]: Nice catch. It should gather everything up until space is detected.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py [hunk 17]
```diff
@@ -1928,26 +1932,55 @@ def nodeset_map(self, hostnames: list):
...
+        return Job(id=job_id, duration=duration, name=name, job_state=job_state, required_nodes=required_nodes)
+
     @lru_cache
-    def job(self, job_id: int) -> Optional[Job]:
-        jobInfo = run(f"{self.scontrol} show jobid {job_id}", check=False).stdout.rstrip()
-        if not jobInfo:
-            return None
+    def get_jobs(self) -> List[Job]:
+        res = run(f"{self.scontrol} show jobs", timeout=30)
+        all_jobs = [x for x in res.stdout.split("\n\n")[:-1]]
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/util.py:1973]: nit.

```py
return [self._parse_job_info(b) for b in res.stdout.split("\n\n")[:-1]]
```

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py [hunk 18]
```diff
@@ -199,16 +200,63 @@ def test_nodeset_reservation_err(nodeset, err):
...
+            TimeLimit=02:00:00
+            JobName=myjob
+            JobState=PENDING
+            ReqNodeList=node-[1-10]""",
+            util.Job(
+                id=123,
+                duration=timedelta(days=0, hours=2, minutes=0, seconds=0),
+                name="myjob",
+                job_state="PENDING",
+                required_nodes="node-[1-10]"
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py:230]: Please add, or update test that involves long `ReqNodeList`, e.g. `node-[1-10],grob-pop-[2,1,44-77]`

**Review by mr0re1 — APPROVED** [review]

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py [hunk 19]
```diff
@@ -199,16 +200,63 @@ def test_nodeset_reservation_err(nodeset, err):
...
-    
-    
+
+
+@pytest.mark.parametrize(
+    "job_info,expected_job",
+    [
+        (
+            """JobId=123
+            TimeLimit=02:00:00
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_util.py:221]: Please include test with TimeLimit that uses `days`
